Source/WebCore/editing/cocoa/AttributedString.mm

```diff
@@ -0,0 +1,162 @@
+        return value;
+    }, [] (const URL& value) -> RetainPtr<NSObject> {
+        return (NSURL *)value;
+#if PLATFORM(MAC)
+    }, [] (const RetainPtr<NSFont>& value) -> RetainPtr<NSObject> {
+        return value;
+    }, [] (const RetainPtr<NSColor>& value) -> RetainPtr<NSObject> {
+        return value;
+    }, [] (const RetainPtr<NSTextAttachment>& value) -> RetainPtr<NSObject> {
+        return value;
+#else
+    }, [] (const RetainPtr<UIFont>& value) -> RetainPtr<NSObject> {
+        return value;
+    }, [] (const RetainPtr<UIColor>& value) -> RetainPtr<NSObject> {
+        return value;
+#endif
+    });
+}
+
+static RetainPtr<NSDictionary> toNSDictionary(const HashMap<String, AttributedString::AttributeValue>& map)
+{
+    auto result = adoptNS([[NSMutableDictionary alloc] initWithCapacity:map.size()]);
+    for (auto& pair : map)
+        [result setObject:toNSObject(pair.value).get() forKey:(NSString *)pair.key];
+    return result;
+}
+
+RetainPtr<NSDictionary> AttributedString::documentAttributesAsNSDictionary() const
+{
+    return toNSDictionary(documentAttributes);
+}
+
+RetainPtr<NSAttributedString> AttributedString::nsAttributedString() const
+{
+    auto result = adoptNS([[NSMutableAttributedString alloc] initWithString:(NSString *)string]);
+    for (auto& pair : attributes) {
+        auto& map = pair.second;
+        auto& range = pair.first;
+        [result addAttributes:toNSDictionary(map).get() range:NSMakeRange(range.location, range.length)];
+    }
+    return result;
+}
+
+static std::optional<AttributedString::AttributeValue> extractValue(id value)
+{
+    if (auto* number = dynamic_objc_cast<NSNumber>(value))
+        return { { { number.doubleValue } } };
+    if (auto* string = dynamic_objc_cast<NSString>(value))
+        return { { { String { string } } } };
+    if (auto* url = dynamic_objc_cast<NSURL>(value))
+        return { { { URL { url } } } };
+#if PLATFORM(MAC)
+    if (auto* paragraphStyle = dynamic_objc_cast<NSParagraphStyle>(value))
+        return { { { RetainPtr { paragraphStyle } } } };
+    if (auto* font = dynamic_objc_cast<NSFont>(value))
+        return { { { RetainPtr { font } } } };
+    if (auto* color = dynamic_objc_cast<NSColor>(value))
+        return { { { RetainPtr { color } } } };
+    if (auto* textAttachment = dynamic_objc_cast<NSTextAttachment>(value))
+        return { { { RetainPtr { textAttachment } } } };
+#else
+    if ([value isKindOfClass:PAL::getNSParagraphStyleClass()])
+        return { { { RetainPtr { (NSParagraphStyle *)value } } } };
+    if ([value isKindOfClass:PAL::getUIFontClass()])
+        return { { { RetainPtr { (UIFont *)value } } } };
+    if ([value isKindOfClass:PAL::getUIColorClass()])
+        return { { { RetainPtr { (UIColor *)value } } } };
+#endif
+    if (value) {
+        LOG_ERROR("NSAttributedString extraction failed for class <%@>", NSStringFromClass([value class]));
```

Comment:
This should be a `RELEASE_LOG_ERROR` so we can capture these events in sysdiagnose. Otherwise this will only show up in local builds.

Reply:
I like it.